miniDAS/format.py

```diff
@@ -0,0 +1,161 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import asdict, dataclass, field
+from pathlib import Path
+from typing import Any, Literal
+
+import h5py
+import numpy as np
+
+DataUnit = Literal["m/m", "cm/m", "mm/m"]
+ScaledUnit = Literal["ue/m"]
+_ALLOWED_DTYPES = (np.int32, np.float32)
+
+
+class OutOfBoundsError(BaseException):
+    ...
+
+
+@dataclass
+class Meta:
+    data_unit: DataUnit
+    start_time_ns: int
+    scale_factor: float
+    units_after_scaling: ScaledUnit
+    sampling_rate: int
+    gauge_length: float
+    latitudes: list[float]
+    longitudes: list[float]
+    elevations: list[float]
+    end_time_ns: int = 0
+
+    @property
+    def deltat(self) -> float:
+        return 1.0 / self.sampling_rate
+
+    @property
+    def start_time(self) -> float:
+        return self.start_time_ns / 1e9
+
+    @property
+    def end_time(self) -> float:
+        return self.end_time_ns / 1e9
+
+
+class miniDAS:
+    _file: h5py.File | None
+
+    version: int = 1
+    dataset: h5py.Dataset | np.ndarray
+    meta: Meta
+
+    def __init__(self, dataset: h5py.Dataset | np.ndarray, meta: Meta):
+        self.dataset = dataset
+        self._file = None
+
+        if len(meta.latitudes) != len(meta.longitudes) != len(meta.elevations):
+            raise AttributeError(
+                "Shape mismatch between latitudes, longitudes and elevation."
+            )
+        if len(meta.latitudes) != self.nchannels:
+            raise AttributeError("Missmatch between number of channels and data shape.")
+
+        meta.end_time_ns = int(meta.start_time_ns + meta.deltat * self.nsamples * 1e9)
+        self.meta = meta
+
+    def write(
+        self, file: Path | str, compress: bool = True, force: bool = False
+    ) -> None:
+        file = Path(file)
+        if file.exists() and not force:
+            raise OSError(f"{file} already exists, use force to overwrite")
+
+        if (
+            isinstance(self.dataset, np.ndarray)
+            and self.dataset.dtype not in _ALLOWED_DTYPES
+        ):
+            raise TypeError(f"Data has invalid dtype {self.dataset.dtype}")
+
+        with h5py.File(file, "w") as container:
+            logging.debug("Writing miniDAS to %s", file)
+            dataset = container.create_dataset(
+                "miniDAS",
+                chunks=True,
+                data=self.dataset,
+                compression="lzf" if compress else False,
+            )
+
+            # dataset.dims[0].label = "channel"
+            # dataset.dims[1].label = "time"
+
+            dataset.attrs["version"] = self.version
+            dataset.attrs.update(asdict(self.meta))
```

Comment:
One of the strength of the original miniDAS was the custom header. 
This is a major drawback in many formats like SEGY, SEG2, miniSEED,... is the inflexibility to allow keeping custom information in the header. Sometimes you just want stuff like project name, customer, explosive size, field crew, well coordinates, whatever stored for eternity... The approach of miniDAS with a dedicated h5-dataset, with potential sub-sets (to be read as dictionary) addresses that nicely , in my humble opinion